eZ/Publish/Core/Repository/ComparingService.php

```diff
@@ -0,0 +1,101 @@
+use eZ\Publish\SPI\Persistence\Content\Type;
+use eZ\Publish\SPI\Persistence\Content\Type\Handler as ContentTypeHandler;
+
+class ComparingService implements ComparingServiceInterface
+{
+    /** @var \eZ\Publish\SPI\Persistence\Content\Handler */
+    private $contentHandler;
+
+    /** @var \eZ\Publish\SPI\Persistence\Content\Type\Handler */
+    private $contentTypeHandler;
+
+    /** @var \eZ\Publish\Core\Compare\FieldRegistry */
+    private $fieldRegistry;
+
+    /** @var \eZ\Publish\Core\Compare\CompareEngineRegistry */
+    private $comparatorEngineRegistry;
+
+    /** @var \eZ\Publish\Core\Repository\Helper\ContentTypeDomainMapper */
+    private $contentTypeDomainMapper;
+
+    public function __construct(
+        ContentHandler $contentHandler,
+        ContentTypeHandler $contentTypeHandler,
+        FieldRegistry $fieldRegistry,
+        CompareEngineRegistry $comparatorEngineRegistry,
+        ContentTypeDomainMapper $contentTypeDomainMapper
+    ) {
+        $this->contentHandler = $contentHandler;
+        $this->contentTypeHandler = $contentTypeHandler;
+        $this->fieldRegistry = $fieldRegistry;
+        $this->comparatorEngineRegistry = $comparatorEngineRegistry;
+        $this->contentTypeDomainMapper = $contentTypeDomainMapper;
+    }
+
+    public function compareVersions(
+        VersionInfo $versionA,
+        VersionInfo $versionB,
+        ?string $languageCode = null
+    ): VersionDiff {
+        $content = $this->contentHandler->load($versionA->getContentInfo()->id, $versionA->versionNo);
+        $contentToCompare = $this->contentHandler->load($versionB->getContentInfo()->id, $versionB->versionNo);
+        $fieldsDiff = [];
+        foreach ($content->fields as $field) {
+            $comparableField = $this->fieldRegistry->getType($field->type);
+            $matchingField = $this->getMatchingField($field, $contentToCompare);
+            $fieldDefinition = $this->contentTypeHandler->getFieldDefinition($field->id, Type::STATUS_DEFINED);
+            $dataA = $comparableField->getDataToCompare(
+                $field->value
+            );
+            $dataB = $comparableField->getDataToCompare(
+                $matchingField->value
+            );
+            $diffs = [];
+            foreach ($dataA as $name => $fieldAData) {
+                $engine = $this->comparatorEngineRegistry->getEngine(get_class($fieldAData));
+
+                $diffs[$name] = $engine->compareFieldsData($fieldAData, $dataB[$name]);
+            }
+            $fieldsDiff[$field->fieldDefinitionId] = new FieldDiff(
+                $this->contentTypeDomainMapper->buildFieldDefinitionDomainObject(
+                    $fieldDefinition,
+                    $languageCode ?? $versionA->initialLanguageCode
+                ),
+                $diffs
+            );
+        }
+
+        return new VersionDiff($fieldsDiff);
+    }
+
+    private function getMatchingField(
```

Comment:
What if ContentType has changed between versions? For example when one of the field we removed. Is this situation covered?

Reply:
Content object is always built on the current state of `ContentType`